bittide/src/Bittide/ClockControl/StabilityChecker.hs

```diff
@@ -0,0 +1,68 @@
+-- SPDX-FileCopyrightText: 2022 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE GADTs #-}
+module Bittide.ClockControl.StabilityChecker where
+
+import Clash.Prelude
+import Data.Maybe
+
+
+-- | Strategies to be used with `stabilityChecker`.
+data StabilityStrategy where
+  -- | Remain within a specified @margin@ of a target buffer occupancy for at least
+  -- @cyclesStable@ number of cycles. See 'checkTargetDeviation' for implementation.
+  TargetDeviation ::
+    ( 1 <= cyclesStable) =>
+    SNat margin ->
+    SNat cyclesStable ->
+    StabilityStrategy
+
+-- | Checks whether the @Signal@ of buffer occupancies from an elastic buffer is "stable".
+-- Supports different 'StabilityStrategy's.
+stabilityChecker ::
+  forall dom n .
+  (HiddenClockResetEnable dom, KnownNat n) =>
+  -- | Strategy used to determine whether the @Signal@ of buffer occupancies is "stable".
+  StabilityStrategy ->
+  -- | Incoming buffer occupancy.
+  Signal dom (Unsigned n) ->
+  -- | Stability indicator.
+  Signal dom Bool
+stabilityChecker = \case
+  TargetDeviation maxDev cyclesStable -> checkTargetDeviation maxDev cyclesStable
+
+-- | Checks if the @Signal@ of buffer occupancies from an elastic buffer is stable.
+-- The @Signal@ is considered stable if it stays within a @margin@ of the target buffer
+-- occupancy for @cyclesStable@ number of cycles. The target is set equal to the
+-- current buffer occupancy after a reset and otherwise remains unchanged.
+checkTargetDeviation ::
+  forall dom margin cyclesStable n .
+  (HiddenClockResetEnable dom, 1 <= cyclesStable, KnownNat n) =>
+  -- | Maximum number of elements the incoming buffer occupancy is allowed to deviate
+  -- from the current @target@ for it to be considered "stable".
+  SNat margin ->
+  -- | Minimum number of clock cycles the incoming buffer occupancy must remain within the
+  -- @margin@ for it to be considered "stable".
+  SNat cyclesStable ->
+  -- | Incoming buffer occupancy.
+  Signal dom (Unsigned n) ->
+  -- | Stability indicator.
+  Signal dom Bool
+checkTargetDeviation SNat SNat = mealy go Nothing
```

Comment:
If you start with `(0,0)` you can get rid of the `Maybe`/`fromJust`. The state machine will immediately jump to `(0, target)`.

Reply:
Perhaps we should just initialize the target at half the range of `Unsigned n`.

Setting it to `(0,0)` would always initialize the target at `0`.